src/class/proto_methods.rs

```diff
@@ -0,0 +1,152 @@
+    fn registory() -> &'static PyProtoRegistry;
+}
+
+impl<T: HasProtoRegistry> PyProtoMethods for T {
+    fn async_methods() -> Option<NonNull<PyAsyncMethods>> {
+        NonNull::new(Self::registory().async_methods.load(Ordering::SeqCst))
+    }
+    fn basic_methods() -> Option<NonNull<PyObjectMethods>> {
+        NonNull::new(Self::registory().basic_methods.load(Ordering::SeqCst))
+    }
+    fn buffer_methods() -> Option<NonNull<PyBufferProcs>> {
+        NonNull::new(Self::registory().buffer_methods.load(Ordering::SeqCst))
+    }
+    fn descr_methods() -> Option<NonNull<PyDescrMethods>> {
+        NonNull::new(Self::registory().descr_methods.load(Ordering::SeqCst))
+    }
+    fn gc_methods() -> Option<NonNull<PyGCMethods>> {
+        NonNull::new(Self::registory().gc_methods.load(Ordering::SeqCst))
+    }
+    fn mapping_methods() -> Option<NonNull<PyMappingMethods>> {
+        NonNull::new(Self::registory().mapping_methods.load(Ordering::SeqCst))
+    }
+    fn number_methods() -> Option<NonNull<PyNumberMethods>> {
+        NonNull::new(Self::registory().number_methods.load(Ordering::SeqCst))
+    }
+    fn iter_methods() -> Option<NonNull<PyIterMethods>> {
+        NonNull::new(Self::registory().iter_methods.load(Ordering::SeqCst))
+    }
+    fn sequence_methods() -> Option<NonNull<PySequenceMethods>> {
+        NonNull::new(Self::registory().sequence_methods.load(Ordering::SeqCst))
+    }
+}
+
+#[doc(hidden)]
+pub struct PyProtoRegistry {
+    /// Async protocols.
+    async_methods: AtomicPtr<PyAsyncMethods>,
+    /// Basic protocols.
+    basic_methods: AtomicPtr<PyObjectMethods>,
+    /// Buffer protocols.
+    buffer_methods: AtomicPtr<PyBufferProcs>,
+    /// Descr pProtocols.
+    descr_methods: AtomicPtr<PyDescrMethods>,
+    /// GC protocols.
+    gc_methods: AtomicPtr<PyGCMethods>,
+    /// Mapping protocols.
+    mapping_methods: AtomicPtr<PyMappingMethods>,
+    /// Number protocols.
+    number_methods: AtomicPtr<PyNumberMethods>,
+    /// Iterator protocols.
+    iter_methods: AtomicPtr<PyIterMethods>,
+    /// Sequence protocols.
+    sequence_methods: AtomicPtr<PySequenceMethods>,
+}
+
+impl PyProtoRegistry {
+    pub const fn new() -> Self {
+        PyProtoRegistry {
+            async_methods: AtomicPtr::new(ptr::null_mut()),
+            basic_methods: AtomicPtr::new(ptr::null_mut()),
+            buffer_methods: AtomicPtr::new(ptr::null_mut()),
+            descr_methods: AtomicPtr::new(ptr::null_mut()),
+            gc_methods: AtomicPtr::new(ptr::null_mut()),
+            mapping_methods: AtomicPtr::new(ptr::null_mut()),
+            number_methods: AtomicPtr::new(ptr::null_mut()),
+            iter_methods: AtomicPtr::new(ptr::null_mut()),
+            sequence_methods: AtomicPtr::new(ptr::null_mut()),
+        }
+    }
+    pub fn set_async_methods(&self, methods: PyAsyncMethods) {
+        self.async_methods
+            .store(Box::into_raw(Box::new(methods)), Ordering::SeqCst)
```

Comment:
Yes, but the current code does the same. See https://github.com/PyO3/pyo3/blob/v0.10.1/src/pyclass.rs#L151.
The difference is we cause memory leak for all protocols(e.g., `iter` and `basic`).
We can avoid `Box::new` for those methods by changing the `pub struct PyProtoRegistry` like the below:
```rust
pub struct PyProtoRegistry {
    /// Protocols that needs specific table types
    async_methods: AtomicPtr<ffi::PyAsyncMethods>,
    buffer_methods: AtomicPtr<ffi::PyBufferProcs>,
    mapping_methods: AtomicPtr<ffi::PyMappingMethods>,
    number_methods: AtomicPtr<ffi::PyNumberMethods>,
    sequence_methods: AtomicPtr<ffi::PySequenceMethods>,
    /// Other methods.
    /// AtomicUsize is used instead of `Atomic<Option<ffi::func>>` or so because
    /// we can't use function pointer in const function.
    tp_str: AtomicUsize,
    tp_repr: AtomicUsize.
    ....
}
```
.

I chose this design first but abandoned it due to its ugliness.

Reply:
Cool, I think your design choice is fine. Can we maybe just add a docstring for now to warn calling these twice will leak?